### PR TITLE
#825: Allow passing explicit connection to dataset API methods.

### PR DESCRIPTION
**Pull request by tseaver**
See #825.

### CODE REVIEW

File: gcloud/datastore/dataset.py [hunk 1]
```diff
@@ -38,29 +38,35 @@ def __init__(self, dataset_id, connection=None):
...
         self.connection = connection
 
-    def get(self, keys, missing=None, deferred=None):
+    def get(self, keys, missing=None, deferred=None, connection=None):
         """Proxy to :func:`gcloud.datastore.api.get`.
 
         Passes our ``dataset_id``.
         """
+        if connection is None:
+            connection = self.connection
```
Comment (dhermes) [inline gcloud/datastore/dataset.py]: What if there is a current transaction going on?

**Comment by dhermes** [comment]
I agree with the spirit of #825 but I think the `Dataset` was already doing the right thing. Maybe we should just make the `connection` required in the constructor?

**Comment by tseaver** [comment]
ISTM that it would be better to just delete the `connection` attribute altogether from `Dataset`, because the methods it proxies to already DTRT if pass `connection=None`.

**Comment by dhermes** [comment]
If we go the "clients everywhere" route, then `Dataset` needs to hold the same properties as `_DefaultsContainer`.

**Comment by tseaver** [comment]
I'm actually opposed to the "client" notion at this point:  I think the "use a connection as a context manager" bit would work better.  We don't need to weld the project ID together with the connection, since it is only passed in a few cases.

**Comment by tseaver** [comment]
I think @jgeewax was actually in agreement with that notion at the end of the last call, too.

**Comment by jgeewax** [comment]
I'm a fan of the Client pattern that Danny is tossing around -- take a look at #861? Maybe we can discuss it there?

**Comment by dhermes** [comment]
It would be nice to avoid churn if we could.

I think making connections constructable with one-liner factories that also pull in credentials will make things easier (in addition to making them have long-lived contexts). If we need clients, we can go that route, but if we don't we can move forward instead of chasing our tail.

**Comment by tseaver** [comment]
@dhermes a671672 removes the `Dataset.connection` attribute altogether -- I think it makes for a better DX.

**Comment by dhermes** [comment]
@tseaver How does it help? What is better? It certainly doesn't capture the idea of a pre-loaded object. And after the change `Dataset` just becomes a string with a bunch of methods.

**Comment by tseaver** [comment]
The goal of the `Dataset` class was to ease the process of constructing many-many objects which needed to be passed a `dataset_id` in their ctors.  Bundling in yet-another-way-to-spell-a-non-default-connection adds neither clarity for the user, who is likely to be confused about how to signal it, nor in straightforwardness of implementation.

Except for `Batch` and `Transaction` objects, which necessarily wrap a connection object, I would say that *any* `connection` attribute in a user-visible object is a misfeature, across all the APIs.

**Comment by dhermes** [comment]
This really just brings up the question: do all config defaults matter as a bundle or just the some of them.

**Comment by tseaver** [comment]
ISTM we can't move on this until #861 settles out.

**Comment by dhermes** [comment]
Indeed

**Comment by dhermes** [comment]
@tseaver this doesn't exactly fit the pattern of #910. I image you'll be binding clients to make `Transaction` work anyhow?

**Pull request closed without merging**